File: 프로그래머스/3/49191. 순위/순위.py
def solution(n, results):
    # win[a][b] = a가 b를 이김(알려짐/추론됨)
    win = [[False] * (n + 1) for _ in range(n + 1)]

    for a, b in results:
        win[a][b] = True

    # Floyd-Warshall (전이 관계: a>k and k>b => a>b)
    for k in range(1, n + 1):
        for a in range(1, n + 1):
            if not win[a][k]:
                continue
            row_a = win[a]
            row_k = win[k]
            for b in range(1, n + 1):
                if row_k[b]:
                    row_a[b] = True

    answer = 0
    for i in range(1, n + 1):
        known = 0
        for j in range(1, n + 1):
            if i == j:
                continue
            if win[i][j] or win[j][i]:
                known += 1
        if known == n - 1:
            answer += 1

    return answer

File: 프로그래머스/3/49191. 순위/순위.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int n, vector<vector<int>> results) {
    vector<vector<int>> win(n + 1, vector<int>(n + 1, 0)); // win[a][b] = a가 b를 이김

    for (auto &r : results) {
        int a = r[0], b = r[1];
        win[a][b] = 1;
    }

    // Floyd-Warshall: a>k and k>b => a>b
    for (int k = 1; k <= n; k++) {
        for (int a = 1; a <= n; a++) {
            if (!win[a][k]) continue;
            for (int b = 1; b <= n; b++) {
                if (win[k][b]) win[a][b] = 1;
            }
        }
    }

    int answer = 0;
    for (int i = 1; i <= n; i++) {
        int known = 0;
        for (int j = 1; j <= n; j++) {
            if (i == j) continue;
            if (win[i][j] || win[j][i]) known++;
        }
        if (known == n - 1) answer++;
    }

    return answer;
}